hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -3344,7 +3344,8 @@ boolean delete(String src, boolean recursive, boolean logRetryCache)
     getEditLog().logSync();
     logAuditEvent(ret, operationName, src);
     if (toRemovedBlocks != null) {
-      removeBlocks(toRemovedBlocks); // Incremental deletion of blocks
+      blockManager.getMarkedDeleteQueue().add(
+          toRemovedBlocks.getToDeleteList());
```

Comment:
we remove blocks for several other operations as well. For example, truncate, rename, deleteSnapshot. Do we plan to make them delete block asynchronous as well?

Reply:
> Thanks for the work and sharing the flame graph, which makes it easy to validate the change.
> 
> However, I am still not able to understand why the change improves delete performance. The delete op is done in two steps, step 1 acquire lock, collect blocks, release lock. step 2 acquire lock, delete blocks, release lock.
> 
> The change essentially moves the step2 to another thread. IMO, this approach reduces client perceived latency, which is good. But deleting the blocks still requires holding namespace lock. Why does it avoid NN unresponsiveness?
> 
> Is it because instead of releasing the lock after a specified number of blocks, it releases the lock after an absolute time. I can image the absolute time is a better metric because deleting a block does take a variable duration of time, not constant.
> 
> A few minor comments changes requested:

Thanks for your comment, as you have commented, the current modification is just to delete the block asynchronously. The QuotaCount calculation optimization described in jira can reduce the time to collect blocks. I plan to open a new problem to solve it.

Reply:
Thank you for your reminder, I will revise them next.